### PR TITLE
Upgrades to PSymbolic (updated PR)

### PR DESCRIPTION
**Pull request by aman-goel**
Hi,

This is the updated version of PR #427.

Updates include: 
- Removes commits from [fp branch](https://github.com/p-org/P/tree/fp) (through git rebase), including `pred enum` construct, Examples/OOPSLA.
- Adds concretizer printing controlled by verbosity flag
- Adds [README](https://github.com/aman-goel/P/blob/master/Src/PRuntimes/PSymbolicRuntime/README.md) detailing solver installation and usage
- Removes 6_SimpleClientServer from Tutorial
- Improved incremental backtracking and statistics collection to enable parallelism in the future

Original PR:

I have added several upgrades to PSymbolic. Here is a summary of major changes:

- Adds SAT/SMT solver backends to easily choose backends via command-line option
- Adds native expression library to express Boolean functions independent of solver backend
- Adds supporting expressions as And-Inverter-Graphs (AIG) and Functionally-Reduced AIGs via [Java bindings](https://github.com/aman-goel/abc_java_bindings) for [ABC tool](https://github.com/berkeley-abc/abc)
- Adds coverage reporting based on choices covered/remaining
- Adds random mode for choice selection
- Adds incremental backtracking to jump directly to the backtrack point
- Adds printing protocol + schedule state in a file and resuming run by reading from file
- Adds command-line options for specifying: solver type, expression type, time limit, memory limit, iteration bound, disabling backtracking, disabling randomization, choosing random seed, resuming run from file, and writing run to file

@ankushdesai, @lmpick, it would be great if you can review and let me know of any suggested changes.

--
Aman
@aman-goel

### CODE REVIEW

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/EntryPoint.java [hunk 1]
```diff
@@ -3,43 +3,157 @@
...
+import psymbolic.valuesummary.solvers.SolverEngine;
 
 import java.time.Duration;
 import java.time.Instant;
+import java.util.concurrent.*;
 
 public class EntryPoint {
-
     public static Instant start = Instant.now();
+    private static ExecutorService executor;
```
[the remark this answers is not in the document]
Comment (aman-goel) [reply Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/EntryPoint.java:17]: This relates to running PSym search with a time limit. Now that PSym has a command-line option `--time-limit <limit-in-seconds>` to specify a time budget for the search, using the ExecutorService (like [here](https://stackoverflow.com/questions/1164301/how-do-i-call-some-blocking-method-with-a-timeout-in-java)) enables running search until search completes or time budget has expired (whatever happens first). In the latter case, we can now throw a TimeoutException and track/print the search result, coverage, and run statistics in a cleaner way.

File: Src/PRuntimes/PSymbolicRuntime/pom.xml [hunk 2]
```diff
@@ -34,7 +120,7 @@
               <artifactId>pjbdd</artifactId>
               <version>1.0.10-9</version>
               <packaging>jar</packaging>
-              <file>lib/pjbdd-v1.0.10-9.jar</file>
+              <file>${pjbdd.jarpath}</file>
```
Comment (ankushdesai) [inline Src/PRuntimes/PSymbolicRuntime/pom.xml:123]: Can I request you to upgrade to the latest version of PJBDD as well. So we all the changes in one PR. thanks!

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java [hunk 3]
```diff
@@ -26,6 +31,74 @@ public class PSymOptions {
...
+                .build();
+        options.addOption(readFromFile);
+
+        // Enable writing the program state to file
+        Option writeToFile = Option.builder("w")
+                .longOpt("write")
+                .desc("Enable writing program state")
+                .numberOfArgs(0)
+                .build();
+        options.addOption(writeToFile);
```
Comment (ankushdesai) [inline Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java:60]: What is the default location for writing the file, can you specify the file to read from as well or it searching in a default location?

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/runtime/Concretizer.java [hunk 4]
```diff
@@ -250,12 +252,13 @@ public static GuardedValue<? extends PValue<?>> concretizePType (Object valueSum
...
     /**
      * Get a list of concrete values for the arguments
+     * @param print Enable printing of concrete value
      * @param pc Guard under which to concretize values
      * @param stop specifies when to stop getting more concrete values
      * @param args arguments
      * @return list of concrete values for arguments
      */ 
-    public static List<GuardedValue<List<Object>>> getConcreteValues(Guard pc, Predicate<Integer> stop, Function<ValueSummary, GuardedValue<?>> concretizer, ValueSummary ... args) {
+    public static List<GuardedValue<List<Object>>> getConcreteValues(boolean print, Guard pc, Predicate<Integer> stop, Function<ValueSummary, GuardedValue<?>> concretizer, ValueSummary ... args) {
```
Comment (ankushdesai) [inline Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/runtime/Concretizer.java:261]: Would it be possible to move this as the last parameter in the function and also control this based on the verbosity input command?

**Review by ankushdesai — CHANGES_REQUESTED** [review]

File: Src/PRuntimes/PSymbolicRuntime/pom.xml (hunk 2 above)
Comment (aman-goel) [reply Src/PRuntimes/PSymbolicRuntime/pom.xml:123]: Updated pjbdd to the latest release from https://www.sosy-lab.org/ivy/org.sosy_lab/pjbdd/

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java (hunk 3 above)
Comment (aman-goel) [reply Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java:60]: When `--write` command-line argument is passed, output files (there can be multiple of them, one each for each backtracking point) are written in the relative `output/` directory, same as where search statistics and trace files are written.

When reading from a file, the file is provided from the command-line as `--read <path-to-file>`, e.g., `--read ../somepath/filename.out`

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/runtime/Concretizer.java (hunk 4 above)
Comment (aman-goel) [reply Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/runtime/Concretizer.java]: I removed the parameter completely, and instead added a static flag in [Concretizer](https://github.com/aman-goel/P/blob/cd486d40c187e6a897d3ec14e0b8e0c42da452c2/Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/runtime/Concretizer.java#L17) that is controlled through [verbosity flag](https://github.com/aman-goel/P/blob/cd486d40c187e6a897d3ec14e0b8e0c42da452c2/Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/EntryPoint.java#L81) to trigger verbose printing in all concrete enumeration scenarios.

File: Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java (hunk 3 above)
Comment (ankushdesai) [reply Src/PRuntimes/PSymbolicRuntime/src/main/java/psymbolic/commandline/PSymOptions.java:60]: Thanks!

**Review by ankushdesai — APPROVED** [review]

**Comment by ankushdesai** [comment]
Thanks, this looks great!

**Pull request merged**